File: geekbang/algorithm/source-code/1/lc20 有效的括号.cpp
class Solution {
public:
    bool isValid(string s) {
        for (char ch : s) {
            if (ch == '(' || ch == '[' || ch == '{') a.push(ch);
            else {
                if (a.empty()) return false;
                if (ch == ')' && a.top() != '(') return false;
                if (ch == ']' && a.top() != '[') return false;
                if (ch == '}' && a.top() != '{') return false;
                a.pop();
            }
        }
        return a.empty();
    }

private:
    stack<char> a;
};

File: geekbang/algorithm/source-code/1/lc20 有效的括号.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> a = new Stack<>();
        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '[' || ch == '{') a.push(ch);
            else {
                if (a.empty()) return false;
                if (ch == ')' && a.peek() != '(') return false;
                if (ch == ']' && a.peek() != '[') return false;
                if (ch == '}' && a.peek() != '{') return false;
                a.pop();
            }
        }
        return a.empty();
    }
}